src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+    return proj_to_loop;
   }
-#endif
 
-  C->print_method(PHASE_BEFORE_LOOP_UNSWITCHING, 4, head);
+ public:
+  IfNode* unswitch_candidate() const {
+    return _unswitch_candidate;
+  }
 
-  // Need to revert back to normal loop
-  if (head->is_CountedLoop() && !head->as_CountedLoop()->is_normal_loop()) {
-    head->as_CountedLoop()->set_normal_loop();
+  IfNode* selector() const {
+    return _selector;
   }
 
-  IfNode* invar_iff = create_slow_version_of_loop(loop, old_new, unswitch_iff, CloneIncludesStripMined);
-  ProjNode* proj_true = invar_iff->proj_out(1);
-  verify_fast_loop(head, proj_true);
+  IfTrueNode* true_path_loop_proj() const {
+    return _true_path_loop_proj;
+  }
 
-  // Increment unswitch count
-  LoopNode* head_clone = old_new[head->_idx]->as_Loop();
-  int nct = head->unswitch_count() + 1;
-  head->set_unswitch_count(nct);
-  head_clone->set_unswitch_count(nct);
+  IfFalseNode* false_path_loop_proj() const {
+    return _false_path_loop_proj;
+  }
+};
 
-  // Hoist invariant casts out of each loop to the appropriate
-  // control projection.
+// See comments below file header for more information about Loop Unswitching.
+void PhaseIdealLoop::do_unswitching(IdealLoopTree* loop, Node_List& old_new) {
+  assert(LoopUnswitching, "LoopUnswitching must be enabled");
 
-  Node_List worklist;
-  for (DUIterator_Fast imax, i = unswitch_iff->fast_outs(imax); i < imax; i++) {
-    ProjNode* proj= unswitch_iff->fast_out(i)->as_Proj();
-    // Copy to a worklist for easier manipulation
-    for (DUIterator_Fast jmax, j = proj->fast_outs(jmax); j < jmax; j++) {
-      Node* use = proj->fast_out(j);
-      if (use->Opcode() == Op_CheckCastPP && loop->is_invariant(use->in(1))) {
-        worklist.push(use);
-      }
-    }
-    ProjNode* invar_proj = invar_iff->proj_out(proj->_con)->as_Proj();
-    while (worklist.size() > 0) {
-      Node* use = worklist.pop();
-      Node* nuse = use->clone();
-      nuse->set_req(0, invar_proj);
-      _igvn.replace_input_of(use, 1, nuse);
-      register_new_node(nuse, invar_proj);
-      // Same for the clone
-      Node* use_clone = old_new[use->_idx];
-      _igvn.replace_input_of(use_clone, 1, nuse);
-    }
+  LoopNode* original_head = loop->_head->as_Loop();
+  if (has_control_dependencies_from_predicates(original_head)) {
+    NOT_PRODUCT(trace_loop_unswitching_impossible(original_head);)
+    return;
   }
 
-  // Hardwire the control paths in the loops into if(true) and if(false)
-  _igvn.rehash_node_delayed(unswitch_iff);
-  dominated_by(proj_true->as_IfProj(), unswitch_iff);
+  NOT_PRODUCT(trace_loop_unswitching_count(loop, original_head);)
+  C->print_method(PHASE_BEFORE_LOOP_UNSWITCHING, 4, original_head);
 
-  IfNode* unswitch_iff_clone = old_new[unswitch_iff->_idx]->as_If();
-  _igvn.rehash_node_delayed(unswitch_iff_clone);
-  ProjNode* proj_false = invar_iff->proj_out(0);
-  dominated_by(proj_false->as_IfProj(), unswitch_iff_clone);
```

Comment:
Extracted to `remove_unswitch_candidate_from_loops()` and put inside `create_unswitched_loop_versions()`.